build-tools/xaprepare/xaprepare/Steps/Step_PrepareBundle.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.IO;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Xamarin.Android.Prepare
+{
+	partial class Step_PrepareBundle : StepWithDownloadProgress
+	{
+		static Uri BundleUriPrefix   => Configurables.Urls.Bundle_XABundleDownloadPrefix;
+		static string BundleFileName => Configurables.Paths.XABundleFileName;
+		static bool osSupportsMonoBuild;
+
+		public Step_PrepareBundle ()
+			: base ("Preparing the binary bundle")
+		{
+			InitOS ();
+		}
+
+		protected override async Task<bool> Execute (Context context)
+		{
+			if (context.ForceRuntimesBuild) {
+				if (osSupportsMonoBuild) {
+					Log.InfoLine ("Rebuilding Mono runtimes as requested");
+					return false;
+				}
+
+				Log.InfoLine ($"Forced Mono runtimes rebuild requested but rebuilding on {context.OS.Type} is currently not supported.");
+			}
+
+			string localPackagePath = Path.Combine (Configurables.Paths.Bundle_ArchivePath);
+			if (await Utilities.VerifyArchive (localPackagePath)) {
+				Log.StatusLine ("Xamarin.Android Bundle archive already downloaded and valid");
+			} else {
+
+				var bundleUrl = new Uri (BundleUriPrefix, BundleFileName);
+
+				Log.StatusLine ("Bundle URL: ", $"{bundleUrl}", tailColor: ConsoleColor.Cyan);
+
+				HttpStatusCode status;
+				bool success;
+				ulong size;
+
+				(success, size, status) = await Utilities.GetDownloadSizeWithStatus (bundleUrl);
+				if (!success) {
+					if (status == HttpStatusCode.NotFound) {
+						if (osSupportsMonoBuild)
+							Log.StatusLine ("   not found, will need to rebuild");
+						else
+							Log.ErrorLine ($"   not found, rebuilding on {context.OS.Type} is not currently supported");
+						return false;
+					}
+
+					throw new InvalidOperationException ($"Failed to access bundle at {bundleUrl} (HTTP status: {status})");
```

Comment:
Could this use the error message we have here: https://github.com/xamarin/xamarin-android/blob/46b4c53dc69e41db905bc4b62393d9c8d064d425/build-tools/download-bundle/download-bundle.targets#L32

I think it explains to external contributors a little better.

There might be some other places in here that would hit, this too?